seqr/views/utils/dataset_utils.py

```diff
@@ -111,40 +111,40 @@ def _find_or_create_missing_sample_records(
         samples,
         projects,
         user,
-        sample_ids,
+        sample_count,
+        get_individual_sample_key,
+        remaining_sample_keys,
         raise_no_match_error=False,
         raise_unmatched_error_template=None,
         create_active=False,
         get_individual_sample_lookup=_get_individual_sample_lookup,
-        get_individual_sample_key=None,
         **kwargs
 ):
     samples = list(samples)
-    remaining_sample_ids = set(sample_ids) - {sample.sample_id for sample in samples}
     matched_individual_ids = {sample.individual_id for sample in samples}
-    if len(remaining_sample_ids) > 0:
+    if len(remaining_sample_keys) > 0:
         remaining_individuals_dict = get_individual_sample_lookup(
             Individual.objects.filter(family__project__in=projects).exclude(id__in=matched_individual_ids)
         )
 
         # find Individual records with exactly-matching individual_ids
         sample_id_to_individual_record = {}
-        for sample_id in remaining_sample_ids:
-            individual_key = get_individual_sample_key(sample_id) if get_individual_sample_key else sample_id
+        for sample_key in remaining_sample_keys:
+            individual_key = get_individual_sample_key(sample_key)
             if individual_key not in remaining_individuals_dict:
                 continue
-            sample_id_to_individual_record[sample_id] = remaining_individuals_dict[individual_key]
+            sample_id_to_individual_record[sample_key] = remaining_individuals_dict[individual_key]
             del remaining_individuals_dict[individual_key]
 
         logger.debug(str(len(sample_id_to_individual_record)) + " matched individual ids", user)
 
-        remaining_sample_ids -= set(sample_id_to_individual_record.keys())
-        if raise_no_match_error and len(remaining_sample_ids) == len(sample_ids):
+        remaining_sample_keys -= set(sample_id_to_individual_record.keys())
+        if raise_no_match_error and len(remaining_sample_keys) == sample_count:
             raise ValueError(
-                'None of the individuals or samples in the project matched the {} expected sample id(s)'.format(
-                    len(sample_ids)
-                ))
-        if raise_unmatched_error_template and remaining_sample_ids:
+                'None of the individuals or samples in the project matched the {} expected sample id(s)'.format(sample_count))
+        if raise_unmatched_error_template and remaining_sample_keys:
+            remaining_sample_ids = {sample_id for sample_id, _ in remaining_sample_keys}\
+                if isinstance(list(remaining_sample_keys)[0], tuple) else remaining_sample_keys
```

Comment:
Still keep the `remaining_sample_ids` because the RNA has the `raise_unmatched_error_template` option. Move the `remaining_sample_ids` generation into the exception branch to avoid unnecessary computation.

Reply:
Then we should change `raise_unmatched_error_template` to `get_unmatched_error` as a function that takes `sample_keys` as an argument and explicitly write the helper functions differently for search and rna samples